File: leetcode-04/solutions/cpp/0763.cpp
class Solution {
 public:
  vector<int> partitionLabels(string S) {
    vector<int> ans;

    vector<int> map(26);
    for (int i = 0; i < S.length(); i++) map[S[i] - 'a'] = i;

    int start = 0;
    int end = 0;

    for (int i = 0; i < S.length(); i++) {
      end = max(end, map[S[i] - 'a']);
      if (i == end) {
        ans.push_back(end - start + 1);
        start = end + 1;
      }
    }

    return ans;
  }
};

File: leetcode-04/solutions/python3/0763.py
class Solution:
    def partitionLabels(self, S: str) -> List[int]:
        ans = []

        dict = {c: i for i, c in enumerate(S)}

        start = 0
        end = 0

        for i, c in enumerate(S):
            end = max(end, dict[c])
            if i == end:
                ans.append(end - start + 1)
                start = end + 1

        return ans
